draft-ietf-tls-esni.md

```diff
@@ -1235,11 +1228,11 @@ can be used to test the encrypted SNI value of specific ClientHello messages.
 ~~~
 {: #tls-resumption-psk title="Message flow for resumption and PSK"}
 
-ECH mitigates against this attack by requiring servers not mix-and-match
-information from the inner and outer ClientHello. For example, if the server
-accepts the inner ClientHello, it does not validate binders in the outer
-ClientHello. This means that ECH PSKs are used within the HPKE encryption
-envelope.
+ECH mitigates against this attack by (1) prohibiting the "pre_shared_key"
+extension in ClientHelloOuter and (2) requiring servers ignore this extension
+when processing ClientHelloOuter. Thus, if the server accepts the inner
+ClientHello, it only validates binders in the inner ClientHello. This means
+that PSKs are used within the HPKE encryption envelope.
```

Comment:
I think this section might be anchoring too much on PSKs. I don't think it's actually a general attack mitigated by a more general fix.

Back when it was ESNI, the final ClientHello was malleable. An attacker could change some parameters while carrying over the secret server name. If the server interprets the influenced parameters differently based on the secret server name, the attacker can use that as an oracle. The most natural vector is the PSK, because resumptions may be SNI-partitioned. But if you have, say, different ALPN preferences by hostname, that would also equally be a vector.

We've fixed this by:
1. Encrypting the entire ClientHello gives us a convenient way to bundle all parameters together, so it's natural that servers don't look at the outer CH. (The section talks about a server that mixes the two CHs up, which is totally bizarre.)
2. Everything that goes into the final ClientHello is directly or indirectly authenticated by the AEAD. (OuterExtensions.inner_digest or #323.)

I'll take a stab at a replacement for this section...

Reply:
> (The section talks about a server that mixes the two CHs up, which is totally bizarre.)

IMO, this is why we kept the section. I agree that it's bizarre, but not much of a stretch from the original attack vector. I'd like to keep some mention of this mix-and-match problem for historical reasons, though I'm definitely happy to look at simpler text!

Reply:
Dunno if it's simpler, but it's hopefully more general. I think the description of the mitigation should touch on OuterExtensions.inner_digest's role in authenticating the inner ClientHello, which would have answered #263. That is, if we allowed you to compress pre_shared_keys for whatever reason, this attack would still be impossible because of that hash.

(Not that you'd want to compress pre_shared_key since a compressed is public and leaking the actual ticket gives you a different kind of oracle. That falls under "things that depend on secret things are also secret".)

Reply:
https://github.com/tlswg/draft-ietf-tls-esni/pull/327